### ISSUE TITLE
Traced Faster RCNN segfaults during forward in C++

### ISSUE THREAD

**Issue by BynaryCobweb** [pytorch/pytorch#53503]
## 🐛 Bug

When calling `forward` on Faster RCNN model from C++ (libtorch), the first `forward` call works fine, and the second `forward` call triggers a segfault.

## To Reproduce

- First convert Faster RCNN model to JIT. I used this script
```python
import torch
import torchvision.models as M

class DetectionModel(torch.nn.Module):
    def __init__(self, model):
        super(DetectionModel, self).__init__()
        self.model = model

    def forward(self, x, bboxes = None, labels = None):
        # type: (Tensor, Optional[Tensor], Optional[Tensor]) -> Tuple[Dict[str, Tensor], List[Dict[str, Tensor]]]
        """
        Input format: one tensor of dimensions [batch size, channel count, width, height]
        """
        l_x = [x[i] for i in range(x.shape[0])]
        if self.training:
            assert bboxes is not None
            assert labels is not None
            l_targs = [{"boxes": bboxes[i], "labels": labels[i]} for i in range(x.shape[0])]
            losses, predictions = self.model(l_x, l_targs)
        else:
            losses, predictions = self.model(l_x)

        return losses, predictions

model = M.detection.fasterrcnn_resnet50_fpn()
detect_model = DetectionModel(model)
detect_model.train()
script_module = torch.jit.script(detect_model)

script_module.save("fasterrcnn.pt")
```
I had to change the input format because I could not find a way to create a List<c10::IValue> from C++ ([I asked on the forum](https://discuss.pytorch.org/t/how-to-create-a-c10-list-ivalue/112261) but got no reply).

I also needed to make a slight change in torchvision for the scripted model to work in training mode. The change is described in this issue: https://github.com/pytorch/pytorch/issues/53015

- Run it twice from C++
```cpp
// minimal_fasterrcnn_fail.cc

#include <iostream>
#include <torch/script.h>
#include <torch/torch.h>

int main(int argc, char **argv) {
  std::string device = "cuda";

  torch::jit::script::Module jit_model = torch::jit::load("fasterrcnn.pt");
  jit_model.train();
  jit_model.to(device);

  std::vector<c10::IValue> inputs;
  inputs.push_back(torch::rand({1, 3, 224, 224}).cuda());
  inputs.push_back(torch::zeros({1, 1, 4}, torch::kLong).cuda());
  inputs.push_back(torch::full({1, 1}, 1, torch::kLong).cuda());
  inputs[1].toTensor()[0][0][2] = 200;
  inputs[1].toTensor()[0][0][3] = 200;

  std::cout << "first time works well: " << jit_model.forward(inputs)
            << std::endl;
  std::cout << "second time segfaults: " << jit_model.forward(inputs)
            << std::endl;
}
```
We have to link with Torchvison for the example to work (because we need the Faster RCNN operators), e.g. in CMake:
```cmake
find_package(Torch REQUIRED)
find_package(TorchVision REQUIRED)

add_executable(fasterrcnn_segfault
    minimal_fasterrcnn_fail.cc
)
target_link_libraries(fasterrcnn_segfault "${TORCH_LIBRARIES}")
target_link_libraries(fasterrcnn_segfault -Wl,--no-as-needed TorchVision::TorchVision)
```

Output:
```
first time works well: [W faster_rcnn.py:95] Warning: RCNN always returns a (Losses, Detections) tuple in scripting (function )
({loss_classifier: 4.52815
[ CUDAFloatType{} ], loss_box_reg: 1.71533e-05
[ CUDAFloatType{} ], loss_objectness: 0.694499
[ CUDAFloatType{} ], loss_rpn_box_reg: 0.0215857
[ CUDAFloatType{} ]}, [])
second time segfaults: Segmentation fault (core dumped)
```

## Expected behavior

No segfault

## Environment

```
PyTorch version: 1.7.1
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.18.4

Python version: 3.6 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: TITAN X (Pascal)
GPU 1: GeForce GTX 1080 Ti
GPU 2: GeForce GTX 1080 Ti

Nvidia driver version: 455.23.05
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] torch==1.7.1
[pip3] torchfile==0.1.0
[pip3] torchvision==0.8.2
[pip3] torchviz==0.0.1
```

**Comment by BynaryCobweb** [comment]
Update: I reproduced the bug in python only
```python
import torch
import torchvision.models as M

class DetectionModel(torch.nn.Module):
    def __init__(self, model):
        super(DetectionModel, self).__init__()
        self.model = model

    def forward(self, x, bboxes = None, labels = None):
        # type: (Tensor, Optional[Tensor], Optional[Tensor]) -> Tuple[Dict[str, Tensor], List[Dict[str, Tensor]]]
        """
        Input format: one tensor of dimensions [batch size, channel count, width, height]
        """
        l_x = [x[i] for i in range(x.shape[0])]
        if self.training:
            assert bboxes is not None
            assert labels is not None
            l_targs = [{"boxes": bboxes[i], "labels": labels[i]} for i in range(x.shape[0])]
            losses, predictions = self.model(l_x, l_targs)
        else:
            losses, predictions = self.model(l_x)

        return losses, predictions

model = M.detection.fasterrcnn_resnet50_fpn()
detect_model = DetectionModel(model)
detect_model.train()
script_module = torch.jit.script(detect_model)
script_module.train()
script_module.cuda()

out = script_module(
        torch.rand(1, 3, 224, 224).cuda(),
        torch.Tensor([1, 1, 200, 200]).unsqueeze(0).unsqueeze(0).cuda(),
        torch.full((1,1), 1).cuda().long()
)
print(out)

out = script_module(
        torch.rand(1, 3, 224, 224).cuda(),
        torch.Tensor([[1, 1, 200, 200], [1, 1, 200, 200]]).unsqueeze(0).cuda(),
        torch.full((1,1), 1).long().cuda()
)
print(out)
```
